qualification/packet_link_qualification.proto

```diff
@@ -0,0 +1,209 @@
+  // keeps the results until the corresponding stop request. Each start response
+  // matches with a corresponding unique operation ID. If there is any error
+  // starting the qualification, status is populated with canonical error code
+  // and message to denote the details.
+  rpc StartPacketQualification(StartPacketQualificationRequest)
+      returns (StartPacketQualificationResponse) {}
+
+  // Stops packet based link qualification on a set of interfaces. The caller
+  // uses the operation ID it previously used when starting the operation to
+  // stop it. Note that a qualification operation is considered completed if the
+  // device has a record/history of it. Also note that it is OK to receive a
+  // stop request for an interface which has completed qualification, as long as
+  // the recorded operation ID matches the one specified by the request. Devices
+  // should clean up at stop request. Each stop response matches with a
+  // corresponding unique operation ID supplied in the request. If there is any
+  // error processing the request, status is populated with canonical error
+  // code and message to denote the details.
+  rpc StopPacketQualification(StopPacketQualificationRequest)
+      returns (StopPacketQualificationResponse) {}
+
+  // Gets packet based link qualification results for a set of operation IDs.
+  // Note that the results will be incomplete during the operation. The caller
+  // uses the operation ID it previously used when starting the operation to
+  // query it. The device is expected to keep the results until the
+  // corresponding stop request. Each get response matches with a corresponding
+  // unique operation ID supplied in the request. If there is any error getting
+  // the results, status is populated with canonical error code and message to
+  // denote the details.
+  rpc GetPacketQualificationResult(GetPacketQualificationResultRequest)
+      returns (GetPacketQualificationResultResponse) {}
+}
+
+// States of qualification.
+enum QualificationState {
+  QUALIFICATION_STATE_UNSPECIFIED = 0;
+  QUALIFICATION_STATE_IDLE = 1;       // Not set up.
+  QUALIFICATION_STATE_READY = 2;      // Set up completed, ready to qualify.
+  QUALIFICATION_STATE_RUNNING = 3;    // Qualification underway.
+  QUALIFICATION_STATE_COMPLETED = 4;  // Qualification is complete.
+  QUALIFICATION_STATE_FAILED = 5;     // Qualification failed for some reasons.
+}
+
+message StartPacketQualificationRequest {
+  enum QualificationEnd {
+    END_UNSPECIFIED = 0;
+    // Responsible for generating loopback packets, etc.
+    NEAR_END = 1;
+    // Responsible for looping back packets only.
+    FAR_END = 2;
+  }
+
+  message QualificationConfiguration {
+    // The ID assigned to the given test. The ID value must be unique among the
+    // tests running on this agent.
+    string id = 1;
+
+    // Path to the interface corresponding to the port to test.
+    .gnoi.types.Path interface = 2;
+
```

Comment:
Are we planning to add "packet_size" in different patch?

Reply:
We can add it to this version. I will work with Yushi to add it here.